### PR TITLE
Gutenberg: Blocks warning dialog

### LINKED ISSUE #28339: Gutenberg: add blocks warning in classic editor
If we detect a block in post_content when using the classic calypso editor (see function below), let's show a modal pop, and offer folks a choice of which editor to use. Behavior should be similar to the opt-in modal, when switching back to the Block Editor/Gutenberg. Clicking on revisions, should dispatch the action from:

`import { openPostRevisionsDialog } from 'state/posts/revisions/actions';`

To allow folks to select a revision to view/load.

43929683-dce0728c-9bea-11e8-9fe2-b1214332e8ae

```jsx
 * Returns true if we detect a core Gutenberg comment block
 * @param   {string }   content - html string
 * @returns { boolean } true if we think we found a block
 */
function hasGutenbergBlocks( content ) {
	return !! content && content.indexOf( '<!-- wp:' ) !== -1;
}
```

### PR DESCRIPTION
**Pull request by mmtr**
#### Changes proposed in this Pull Request

Displays a warning dialog when we detect in the Classic editor that the post content contains Gutenberg blocks.

screen shot 2018-11-09 at 13 53 11

* When clicking on "Switch to the new editor", it changes the user preferred editor to Gutenberg and redirects to the Gutenberg editor URL.
* When clicking on "Use the classic editor", it dismisses the dialog.
* When clicking on "document history", it shows the revisions dialog and dismisses the blocks warning dialog.

Fixes #28339

#### Testing instructions

**Existing post: Switch to the new editor**
* Load in the Classic editor a post created in Gutenberg (so it contains Gutenberg blocks)
* Confirm that the "Blocks warning" dialog appears
* Click on "Switch to the new editor"
* Make sure that the preferred `editor_web` changes to `gutenberg`
  * Call `GET https://public-api.wordpress.com/wpcom/v2/sites/:site/gutenberg`
* Check that Calypso redirects to the Gutenberg editor which loads the existing post

**Existing post: Use the classic editor**
* Load in the Classic editor a post created in Gutenberg (so it contains Gutenberg blocks)
* Click on "Use the classic editor"
* Check that the "Blocks warning" dialog is dismissed and the Classic editor remains open
* Make sure that the preferred `editor_web` doesn't change
  * Call `GET https://public-api.wordpress.com/wpcom/v2/sites/:site/gutenberg`
* Confirm that the "Blocks warning" dialog doesn't show up again while continuing editing the post in the Classic editor
* Reload the page (or close and edit again the same post having Gutenberg blocks) and check that the "Blocks warning" dialog appears again

**Existing post: Document history**
* Load in the Classic editor a post created in Gutenberg (so it contains Gutenberg blocks)
* Click on "document history"
* Check that the "Blocks warning" dialog is dismissed, the Classic editor remains open and the "Revisions" dialog is displayed
* Make sure that the preferred `editor_web` doesn't change
  * Call `GET https://public-api.wordpress.com/wpcom/v2/sites/:site/gutenberg`
* Load a previous version and confirm that the "Blocks warning" dialog doesn't show up again while continuing editing the post in the Classic editor

**Existing post without Gutenberg blocks**
* Load in the Classic editor a post created in the Classic editor (so it doesn't contain Gutenberg blocks)
* Confirm that the "Blocks warning" dialog doesn't appears

**New post**
* Start creating a new post using the Classic editor
* Confirm that the "Blocks warning" dialog doesn't appears
* Switch to the HTML mode and enter the following:
```
<!-- wp:paragraph -->
Paragraph
<!-- /wp:paragraph -->
```
* Confirm that the "Blocks warning" dialog appears

### CODE REVIEW

File: client/state/ui/gutenberg-blocks-warning-dialog/actions.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+/** @format */
+
+/**
+ * Internal dependencies
+ */
+import { GUTENBERG_BLOCKS_WARNING_DIALOG_IS_SHOWING } from 'state/action-types';
+
+export function showGutenbergBlocksWarningDialog() {
+	return {
+		type: GUTENBERG_BLOCKS_WARNING_DIALOG_IS_SHOWING,
```
Comment (gwwar) [inline client/state/ui/gutenberg-blocks-warning-dialog/actions.js]: 👍 if we'd like the editor to toggle display of this. I'm pretty open to options here, for example making the dialog self-contained is also okay (just use component state). The latter would require us to have the dialog responsible for either using passed in post_content / fetching itself + using: 

`import { parse } from '@wordpress/blocks';` or a lighter weight check of `!! content && content.indexOf( '<!-- wp:' ) !== -1;`

Feel free to pick whatever feels better.

Comment (mmtr) [reply client/state/ui/gutenberg-blocks-warning-dialog/actions.js]: Thanks for the suggestion!

I definitely like having components self contained, so they can work out of the box by just importing them (btw, is not that the definition of a component? 😅).

I moved the whole the state to the component itself in 77446ccd1e857353f1aa4a8b107cb4fea570a419

File: client/components/dialog/style.scss [hunk 2]
```diff
@@ -73,7 +73,6 @@
 		color: $gray-dark;
 		font-size: 1.375em;
 		font-weight: 600;
-		height: 2em;
```
Comment (mmtr) [inline client/components/dialog/style.scss]: I removed this height so the title of the dialog can be displayed in 2 lines if it doesn't fit the available width (i.e. in mobile devices)

File: client/post-editor/editor-gutenberg-blocks-warning-dialog/index.jsx [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		return {
+			postContent,
+			siteId,
+			gutenbergUrl,
+		};
+	},
+	dispatch => ( {
+		switchToGutenberg: ( siteId, gutenbergUrl ) => {
+			return dispatch( setSelectedEditor( siteId, 'gutenberg', gutenbergUrl ) );
+		},
```
Comment (dmsnell) [inline client/post-editor/editor-gutenberg-blocks-warning-dialog/index.jsx]: the `return` here is unexpected/inconsistent with that from `openPostReviewionsDialog`

is it needed? it doesn't appear like we're using it, maybe go back to parity with the other?

```js
switchToGutenberg: ( siteId, gutenbergUrl ) => dispatch( setSelectedEditor( siteId, 'gutenberg', gutenbergUrl ) ),
```

File: client/post-editor/editor-gutenberg-blocks-warning-dialog/index.jsx [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+		return {
+			isDialogVisible,
+		};
+	}
+
+	onUseClassic = () => {
+		this.setState( {
+			forceClassic: true,
+		} );
+	};
```
Comment (dmsnell) [inline client/post-editor/editor-gutenberg-blocks-warning-dialog/index.jsx]: we typically avoid using `onXYZ` names and prefer instead names that tersely describe what they do, like `useClassicEditor` and `switchToGutenberg`

[coding guidelines](https://wpcalypso.wordpress.com/devdocs/docs/coding-guidelines/javascript.md) check out the section `React Components`

File: client/post-editor/editor-gutenberg-blocks-warning-dialog/index.jsx [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+					) }
+				</p>
+
+				<p>
+					{ translate(
+						'You can also check the {{a}}document history{{/a}} and restore a version of the page from earlier.',
+						{
+							components: {
+								//eslint-disable-next-line jsx-a11y/anchor-is-valid
+								a: <a href="#" onClick={ e => this.onShowDocumentHistory( e ) } />,
```
Comment (dmsnell) [inline client/post-editor/editor-gutenberg-blocks-warning-dialog/index.jsx]: `e` and the arrow-function isn't necessary…I think we can pass `{ this.showDocumentHistory }` instead

File: client/post-editor/editor-gutenberg-blocks-warning-dialog/index.jsx (hunk 3 above)
Comment (mmtr) [reply client/post-editor/editor-gutenberg-blocks-warning-dialog/index.jsx]: Yeah, the `return` is implicit when using the arrow function in one single line, so both were using the `return`.

I decided to explicitly use it here for avoid a very long line (that is split in 2 by eslint), but I guess it's better to have consistency 😉.

Changed in https://github.com/Automattic/wp-calypso/pull/28406/commits/6ce013ba682c0726b4ca3f23d9ebe5db7c0633c7.

File: client/post-editor/editor-gutenberg-blocks-warning-dialog/index.jsx (hunk 4 above)
Comment (mmtr) [reply client/post-editor/editor-gutenberg-blocks-warning-dialog/index.jsx]: Gotcha! I also like that convention, but I don't know why I was using the React convention for props names. 

Changed in 6ce013b.

**Comment by drw158** [comment]
i would suggest you put all of the body text in a single `p` to avoid the weird line spacing between the 2 paragraphs of text. I think placing all of the text in a single paragraph is preferrable.

**Pull request merged**